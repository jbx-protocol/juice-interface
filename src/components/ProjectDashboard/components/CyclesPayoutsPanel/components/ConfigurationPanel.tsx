import { t } from '@lingui/macro'
import React, { ReactNode } from 'react'
import { ConfigurationTable } from './ConfigurationTable'

export type ConfigurationPanelDatum = {
  name: ReactNode
  old?: ReactNode
  new: ReactNode
}

export type ConfigurationPanelTableData = {
  [key: string]: ConfigurationPanelDatum
}

type ConfigurationPanelProps = {
  cycle: ConfigurationPanelTableData
  token: ConfigurationPanelTableData
  otherRules: ConfigurationPanelTableData
  extension: ConfigurationPanelTableData | null
}

export const ConfigurationPanel: React.FC<ConfigurationPanelProps> = ({
  cycle,
  token,
  otherRules,
  extension,
}) => {
  return (
    <div className="flex flex-col gap-8">
      <ConfigurationTable title={t`Cycle`} data={cycle} />
      <ConfigurationTable title={t`Token`} data={token} />
      <ConfigurationTable title={t`Other rules`} data={otherRules} />
      {extension && (
        <ConfigurationTable title={t`Extension`} data={extension} />
      )}
    </div>
  )
}
