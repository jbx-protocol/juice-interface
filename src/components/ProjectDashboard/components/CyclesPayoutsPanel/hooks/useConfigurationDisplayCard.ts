import { t } from '@lingui/macro'
import { useMemo } from 'react'

export const useConfigurationDisplayCard = (type: 'current' | 'upcoming') => {
  const title = useMemo(() => {
    if (type === 'current') {
      return t`Current cycle`
    }
    return t`Upcoming cycle`
  }, [type])

  return {
    title,
  }
}
