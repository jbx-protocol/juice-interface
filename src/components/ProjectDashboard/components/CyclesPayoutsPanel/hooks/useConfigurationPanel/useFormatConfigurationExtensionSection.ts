import { t } from '@lingui/macro'
import { V2V3FundingCycleMetadata } from 'models/v2v3/fundingCycle'
import { useMemo } from 'react'
import { isZeroAddress } from 'utils/address'
import {
  ConfigurationPanelDatum,
  ConfigurationPanelTableData,
} from '../../components/ConfigurationPanel'
import { flagPairToDatum } from '../../utils/flagPairToDatum'
import { pairToDatum } from '../../utils/pairToDatum'

export const useFormatConfigurationExtensionSection = ({
  fundingCycleMetadata,
  upcomingFundingCycleMetadata,
}: {
  fundingCycleMetadata: V2V3FundingCycleMetadata | undefined
  upcomingFundingCycleMetadata?: V2V3FundingCycleMetadata | null
}): ConfigurationPanelTableData | null => {
  const contractDatum: ConfigurationPanelDatum = useMemo(() => {
    const currentContract = fundingCycleMetadata?.dataSource
    const upcomingContract = upcomingFundingCycleMetadata?.dataSource

    if (upcomingFundingCycleMetadata === null) {
      return pairToDatum(t`Contract`, currentContract, null)
    }

    return pairToDatum(t`Contract`, currentContract, upcomingContract)
  }, [fundingCycleMetadata?.dataSource, upcomingFundingCycleMetadata])

  const useForPaymentsDatum: ConfigurationPanelDatum = useMemo(() => {
    const currentUseForPayments = fundingCycleMetadata?.useDataSourceForPay
    const upcomingUseForPayments =
      upcomingFundingCycleMetadata?.useDataSourceForPay

    if (upcomingFundingCycleMetadata === null) {
      return flagPairToDatum(t`Use for payments`, currentUseForPayments, null)
    }

    return flagPairToDatum(
      t`Use for payments`,
      currentUseForPayments,
      upcomingUseForPayments,
    )
  }, [fundingCycleMetadata?.useDataSourceForPay, upcomingFundingCycleMetadata])

  const useForRedemptionsDatum: ConfigurationPanelDatum = useMemo(() => {
    const currentUseForRedemptions =
      fundingCycleMetadata?.useDataSourceForRedeem
    const upcomingUseForRedemptions =
      upcomingFundingCycleMetadata?.useDataSourceForRedeem

    if (upcomingFundingCycleMetadata === null) {
      return flagPairToDatum(
        t`Use for redemptions`,
        currentUseForRedemptions,
        null,
      )
    }

    return flagPairToDatum(
      t`Use for redemptions`,
      currentUseForRedemptions,
      upcomingUseForRedemptions,
    )
  }, [
    fundingCycleMetadata?.useDataSourceForRedeem,
    upcomingFundingCycleMetadata,
  ])

  const formatted = useMemo(() => {
    return {
      contract: contractDatum,
      useForPayments: useForPaymentsDatum,
      useForRedemptions: useForRedemptionsDatum,
    }
  }, [contractDatum, useForPaymentsDatum, useForRedemptionsDatum])

  if (isZeroAddress(fundingCycleMetadata?.dataSource)) {
    return null
  }

  return formatted
}
