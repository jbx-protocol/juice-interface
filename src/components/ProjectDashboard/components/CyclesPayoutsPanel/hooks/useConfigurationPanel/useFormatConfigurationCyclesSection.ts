import { t } from '@lingui/macro'
import { timeSecondsToDateString } from 'components/ProjectDashboard/utils/timeSecondsToDateString'
import { BigNumber } from 'ethers'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { V2V3FundingCycle } from 'models/v2v3/fundingCycle'
import { useMemo } from 'react'
import { fromWad } from 'utils/format/formatNumber'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'
import { getBallotStrategyByAddress } from 'utils/v2v3/ballotStrategies'
import { MAX_DISTRIBUTION_LIMIT } from 'utils/v2v3/math'
import { ConfigurationPanelDatum } from '../../components/ConfigurationPanel'
import { pairToDatum } from '../../utils/pairToDatum'

export const useFormatConfigurationCyclesSection = ({
  fundingCycle,
  upcomingFundingCycle,
  distributionLimitAmountCurrency,
  upcomingDistributionLimitAmountCurrency,
}: {
  fundingCycle: V2V3FundingCycle | undefined
  upcomingFundingCycle?: V2V3FundingCycle | null
  distributionLimitAmountCurrency:
    | {
        distributionLimit: BigNumber | undefined
        currency: BigNumber | undefined
      }
    | undefined
  upcomingDistributionLimitAmountCurrency?: {
    distributionLimit: BigNumber | undefined
    currency: BigNumber | undefined
  } | null
}) => {
  const durationDatum: ConfigurationPanelDatum = useMemo(() => {
    const formatDuration = (duration: BigNumber | undefined) => {
      if (duration === undefined) return undefined
      if (duration.eq(0)) return t`Not set`
      return timeSecondsToDateString(duration.toNumber(), 'short', 'lower')
    }
    const currentDuration = formatDuration(fundingCycle?.duration)
    if (upcomingFundingCycle === null) {
      return pairToDatum(t`Duration`, currentDuration, null)
    }
    const upcomingDuration = formatDuration(upcomingFundingCycle?.duration)

    return pairToDatum(t`Duration`, currentDuration, upcomingDuration)
  }, [fundingCycle?.duration, upcomingFundingCycle])

  const payoutsDatum: ConfigurationPanelDatum = useMemo(() => {
    const formatCurrency = (currency: BigNumber | undefined) => {
      if (currency === undefined) return undefined
      return currency.toNumber() as V2V3CurrencyOption
    }
    const formatAmountWad = (
      amountWad: BigNumber | undefined,
      currency: V2V3CurrencyOption | undefined,
    ) => {
      if (amountWad === undefined) return undefined
      if (amountWad.eq(MAX_DISTRIBUTION_LIMIT)) return t`Unlimited`
      if (amountWad.eq(0)) return t`Zero (no payouts)`
      return formatCurrencyAmount({
        amount: Number(fromWad(amountWad)),
        currency,
      })
    }
    const { distributionLimit, currency } =
      distributionLimitAmountCurrency ?? {}
    const currentPayout = formatAmountWad(
      distributionLimit,
      formatCurrency(currency),
    )

    if (upcomingDistributionLimitAmountCurrency === null) {
      return pairToDatum(t`Payouts`, currentPayout, null)
    }

    const upcomingDistributionLimit =
      upcomingDistributionLimitAmountCurrency?.distributionLimit !== undefined
        ? upcomingDistributionLimitAmountCurrency.distributionLimit
        : undefined
    const upcomingDistributionLimitCurrency =
      upcomingDistributionLimitAmountCurrency?.currency !== undefined
        ? upcomingDistributionLimitAmountCurrency.currency
        : undefined
    const upcomingPayout = formatAmountWad(
      upcomingDistributionLimit,
      formatCurrency(upcomingDistributionLimitCurrency),
    )

    return pairToDatum(t`Payouts`, currentPayout, upcomingPayout)
  }, [distributionLimitAmountCurrency, upcomingDistributionLimitAmountCurrency])

  const editDeadlineDatum: ConfigurationPanelDatum = useMemo(() => {
    const currentBallotStrategy = fundingCycle?.ballot
      ? getBallotStrategyByAddress(fundingCycle.ballot)
      : undefined
    const current = currentBallotStrategy?.name
    if (upcomingFundingCycle === null) {
      return pairToDatum(t`Edit deadline`, current, null)
    }

    const upcomingBallotStrategy = upcomingFundingCycle?.ballot
      ? getBallotStrategyByAddress(upcomingFundingCycle.ballot)
      : undefined
    const upcoming = upcomingBallotStrategy?.name
    return pairToDatum(t`Edit deadline`, current, upcoming)
  }, [fundingCycle?.ballot, upcomingFundingCycle])

  return useMemo(() => {
    return {
      duration: durationDatum,
      payouts: payoutsDatum,
      editDeadline: editDeadlineDatum,
    }
  }, [durationDatum, editDeadlineDatum, payoutsDatum])
}
