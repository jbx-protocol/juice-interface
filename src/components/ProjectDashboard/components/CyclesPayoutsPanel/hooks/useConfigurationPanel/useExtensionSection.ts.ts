import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'
import { ConfigurationPanelTableData } from '../../components/ConfigurationPanel'
import { useFormatConfigurationExtensionSection } from './useFormatConfigurationExtensionSection'

export const useExtensionSection = (
  type: 'current' | 'upcoming',
): ConfigurationPanelTableData | null => {
  const { projectId } = useProjectMetadata()

  const { fundingCycleMetadata } = useProjectContext()
  const { data: upcomingFundingCycleData } = useProjectUpcomingFundingCycle({
    projectId,
  })
  const [, upcomingFundingCycleMetadata] = upcomingFundingCycleData ?? []

  return useFormatConfigurationExtensionSection({
    fundingCycleMetadata,
    upcomingFundingCycleMetadata,
    ...(type === 'current' && {
      upcomingFundingCycleMetadata: null,
    }),
  })
}
