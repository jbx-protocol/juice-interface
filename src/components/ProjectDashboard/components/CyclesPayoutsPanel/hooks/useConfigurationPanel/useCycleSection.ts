import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'
import { ConfigurationPanelTableData } from '../../components/ConfigurationPanel'
import { useFormatConfigurationCyclesSection } from './useFormatConfigurationCyclesSection'

export const useCycleSection = (
  type: 'current' | 'upcoming',
): ConfigurationPanelTableData => {
  const { projectId } = useProjectMetadata()

  const {
    fundingCycle,
    distributionLimit,
    distributionLimitCurrency,
    primaryETHTerminal,
  } = useProjectContext()
  const { data: upcomingFundingCycleData } = useProjectUpcomingFundingCycle({
    projectId,
  })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const { data: upcomingDistributionLimitData } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })
  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  return useFormatConfigurationCyclesSection({
    fundingCycle,
    distributionLimitAmountCurrency: {
      distributionLimit: distributionLimit,
      currency: distributionLimitCurrency,
    },

    upcomingFundingCycle,
    upcomingDistributionLimitAmountCurrency: {
      distributionLimit: upcomingDistributionLimit,
      currency: upcomingDistributionLimitCurrency,
    },

    // Hide upcoming info from current section.
    ...(type === 'current' && {
      upcomingFundingCycle: null,
      upcomingDistributionLimitAmountCurrency: null,
    }),
  })
}
