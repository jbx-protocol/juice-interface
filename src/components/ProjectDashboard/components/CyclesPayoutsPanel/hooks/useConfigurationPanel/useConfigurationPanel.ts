import { useCycleSection } from './useCycleSection'
import { useOtherRulesSection } from './useOtherRulesSection'
import { useTokenSection } from './useTokenSection'

export const useConfigurationPanel = (type: 'current' | 'upcoming') => {
  const cycle = useCycleSection(type)
  const token = useTokenSection(type)
  const otherRules = useOtherRulesSection(type)

  return {
    cycle,
    token,
    otherRules,
  }
}
