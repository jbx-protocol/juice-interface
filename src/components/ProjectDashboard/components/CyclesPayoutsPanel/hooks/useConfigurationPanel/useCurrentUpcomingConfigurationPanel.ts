import { useCycleSection } from './useCycleSection'
import { useOtherRulesSection } from './useOtherRulesSection'
import { useTokenSection } from './useTokenSection'

export const useCurrentUpcomingConfigurationPanel = (
  type: 'current' | 'upcoming',
) => {
  const cycle = useCycleSection(type)
  const token = useTokenSection(type)
  const otherRules = useOtherRulesSection(type)

  return {
    cycle,
    token,
    otherRules,
  }
}
