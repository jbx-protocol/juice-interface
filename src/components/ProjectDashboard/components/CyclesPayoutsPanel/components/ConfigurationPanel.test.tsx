/**
 * @jest-environment jsdom
 */
import { render } from '@testing-library/react'
import { useConfigurationPanel } from '../hooks/useConfigurationPanel'
import { ConfigurationPanel } from './ConfigurationPanel'

jest.mock('../hooks/useConfigurationPanel', () => ({
  useConfigurationPanel: jest.fn(),
}))

jest.mock('./ConfigurationTable', () => ({
  ConfigurationTable: jest.fn(() => (
    <div data-testid="ConfigurationTable">ConfigurationTable</div>
  )),
}))

const DefaultResponse = {
  cycle: {
    foo: 'foo',
  },
  token: {
    bar: 'bar',
  },
  otherRules: {
    baz: 'baz',
  },
}

const mockUseConfigurationPanel = useConfigurationPanel as jest.Mock

describe('ConfigurationPanel', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    mockUseConfigurationPanel.mockImplementation(() => DefaultResponse)
  })

  it('renders', () => {
    const { container } = render(<ConfigurationPanel type={'current'} />)
    expect(container).toMatchSnapshot()
  })

  it.each(['current', 'upcoming'] as const)(
    'calls useConfigurationPanel with type %p',
    type => {
      render(<ConfigurationPanel type={type} />)
      expect(mockUseConfigurationPanel).toHaveBeenCalledWith(type)
    },
  )
})
