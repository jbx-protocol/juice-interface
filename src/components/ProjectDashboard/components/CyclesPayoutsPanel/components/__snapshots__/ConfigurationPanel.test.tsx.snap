// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigurationPanel renders 1`] = `
<div>
  <div
    class="flex flex-col gap-8"
  >
    <div
      data-testid="ConfigurationTable"
    >
      ConfigurationTable
    </div>
    <div
      data-testid="ConfigurationTable"
    >
      ConfigurationTable
    </div>
    <div
      data-testid="ConfigurationTable"
    >
      ConfigurationTable
    </div>
  </div>
</div>
`;
